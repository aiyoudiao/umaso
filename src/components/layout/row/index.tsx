import type { FC, ReactNode } from 'react'
import React, { createContext } from 'react'
import classNames from 'classnames'
import '../../../theme/src/row.scss'

interface RowType {
  tag?: string
  gutter?: number
  type?: string
  justify?: justifyType | string
  align?: alignType | string
}

enum AlignEnum {
  'top' = 'top',
  'middle' = 'middle',
  'bottom' = 'bottom',
  'stretch' = 'stretch',
}

enum JustifyEnum {
  'start' = 'start',
  'end' = 'end',
  'center' = 'center',
  'spaceAround' = 'space-around',
  'spaceBetween' = 'space-between',
  'spaceEvenly' = 'space-evenly',
}

export type alignType = AlignEnum.top | AlignEnum.middle | AlignEnum.bottom | AlignEnum.stretch
export type justifyType = JustifyEnum.start | JustifyEnum.end | JustifyEnum.center | JustifyEnum.spaceAround | JustifyEnum.spaceBetween | JustifyEnum.spaceEvenly
export interface RowProps extends RowType {
  className?: string
  children?: ReactNode
  style?: React.CSSProperties
}
export interface RowContextState {
  gutter?: number
}
export const RowContext = createContext<RowContextState>({})

export const Row: FC<RowProps> = (props) => {
  const {
    tag = 'div',
    type: typeName,
    align = AlignEnum.top,
    style = {},
    gutter = 0,
    justify = JustifyEnum.start,
    children,
    className: partClassName,
    ...otherList
  } = props

  // 整合样式对象
  let rowStyle: React.CSSProperties = {}
  if (gutter) {
    const margin = `-${gutter / 2}px`
    rowStyle = ({
      marginLeft: margin,
      marginRight: margin,
    })
  }

  // 整合class对象
  const classList: Array<string> = ['el-row']
  if (justify !== 'start')
    classList.push(`is-justify-${justify}`)

  if (align !== 'top')
    classList.push(`is-align-${align}`)

  if (typeName === 'flex')
    classList.push('el-row--flex')

  const className = classNames(partClassName, [...classList].join(' '))
  const rowContext = React.useMemo(() => ({ gutter }), [gutter])
  const RowComponent = React.createElement(RowContext.Provider, { value: rowContext },
    React.createElement(
      tag,
      {
        name: 'el-row',
        className,
        style: { ...rowStyle, ...style },
        ...otherList,
      },
      children,
    ))

  return RowComponent
}
