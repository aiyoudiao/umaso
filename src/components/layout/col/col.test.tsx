import React from 'react'
import type { ReactTestRendererJSON } from 'react-test-renderer'
import renderer from 'react-test-renderer'
import { describe, expect, it } from 'vitest'
import { Row as ElRow } from '../row'
import { Col as ElCol } from './index'

describe('Col.react', () => {
  it('Col snapshot', () => {
    const component = renderer.create(<ElCol></ElCol>)
    expect(component).toMatchSnapshot()
  })
  describe('Col class', () => {
    it ('contain has class .el-col', () => {
      const component = renderer.create(
        <ElCol></ElCol>,
      )
      const { type, props: { className, name } } = component.toJSON() as ReactTestRendererJSON
      expect(type).toBe('div')
      expect(name).toBe('el-col')
      expect(className).toBe('el-col')
    })
    it('contain has class .el-col-{num}', () => {
      const num = 12
      const component = renderer.create(
        <ElCol span={num}></ElCol>,
      )
      const { props: { className } } = component.toJSON() as ReactTestRendererJSON
      expect(className).toContain(`el-col-${num}`)
    })
    it('contain has class .el-col-offset-{num}', () => {
      const num = 8
      const component = renderer.create(
        <ElCol offset={num}></ElCol>,
      )
      const { props: { className } } = component.toJSON() as ReactTestRendererJSON
      expect(className).toContain(`el-col-offset-${num}`)
    })
    it('contain has class .el-col-pull-{num}', () => {
      const num = 6
      const component = renderer.create(
        <ElCol pull={num}></ElCol>,
      )
      const { props: { className } } = component.toJSON() as ReactTestRendererJSON
      expect(className).toContain(`el-col-pull-${num}`)
    })
    it('contain has class .el-col-push-{num}', () => {
      const num = 4
      const component = renderer.create(
        <ElCol push={num}></ElCol>,
      )
      const { props: { className } } = component.toJSON() as ReactTestRendererJSON
      expect(className).toContain(`el-col-push-${num}`)
    })
  })

  it('Col gutter style', () => {
    const count = 10
    const gutter = count * 2
    const component = renderer.create(
      <ElRow gutter={gutter}>
        <ElCol></ElCol>
      </ElRow>,
    )
    const componentInstance = component.root
    const { props: { name, style } } = componentInstance.findByProps({ className: 'el-col' })
    expect(name).toBe('el-col')
    expect(style).toEqual({
      paddingLeft: `${count}px`,
      paddingRight: `${count}px`,
    })
  })

  it('The responsive', () => {
    const component = renderer.create(
      <ElRow gutter={22}>
        <ElCol
          xs={{ span: 22, offset: 2 }}
          sm={{ span: 18, offset: 4 }}
          md={22}
          lg={{ span: 20, offset: 2 }}
          xl={18}
        >
          hello world
        </ElCol>
      </ElRow>,
    )
    const componentInstance = component.root
    const { props: { className, style } } = componentInstance.find(node => node.props.name === 'el-col')
    expect(className).toContain('el-col')
    expect(style).toEqual({
      paddingLeft: '11px',
      paddingRight: '11px',
    })
    expect(className).toContain('el-col-xs-22')
    expect(className).toContain('el-col-xs-offset-2')
    expect(className).toContain('el-col-sm-18')
    expect(className).toContain('el-col-sm-offset-4')
    expect(className).toContain('el-col-md-22')
    expect(className).toContain('el-col-lg-20')
    expect(className).toContain('el-col-lg-offset-2')
    expect(className).toContain('el-col-xl-18')
  })
})
