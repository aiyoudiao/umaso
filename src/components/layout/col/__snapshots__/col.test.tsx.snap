// Vitest Snapshot v1

exports[`Col.react > Col snapshot 1`] = `
<div
  className="el-col"
  name="el-col"
  style={{}}
/>
`;
